fpga_interchange/populate_chip_info.py

```diff
@@ -799,6 +801,658 @@ def print_bel_buckets(cell_bel_mapper):
+        for tile_type_idx in tile_types_with_gnd:
+            tile_type = self.chip_info.tile_types[tile_type_idx]
+
+            gnd_wire_idx = len(tile_type.wire_data)
+            gnd_wire = TileWireInfo()
+            tile_type.wire_data.append(gnd_wire)
+
+            gnd_wire.name = '$GND_WIRE'
+            gnd_wire.site = -1
+
+            tile_type_gnd_wires[tile_type_idx] = gnd_wire_idx
+
+        # Add gnd local wire in each instance to node
+        for tile_idx, tile in enumerate(self.chip_info.tiles):
+            if tile.type not in tile_types_with_gnd:
+                continue
+
+            gnd_wire_idx = tile_type_gnd_wires[tile.type]
+            assert gnd_wire_idx >= len(
+                tile.tile_wire_to_node), (gnd_wire_idx,
+                                          len(tile.tile_wire_to_node),
+                                          len(tile_type.wire_data))
+
+            while gnd_wire_idx > len(tile.tile_wire_to_node):
+                tile.tile_wire_to_node.append(-1)
+
+            tile.tile_wire_to_node.append(gnd_node_idx)
+
+            wire_ref = TileWireRef()
+            wire_ref.tile = tile_idx
+            wire_ref.index = gnd_wire_idx
+
+            gnd_node.tile_wires.append(wire_ref)
+
+        # Create vcc local wire in each type that needs it
+        tile_type_vcc_wires = {}
+        for tile_type_idx in tile_types_with_vcc:
+            tile_type = self.chip_info.tile_types[tile_type_idx]
+
+            vcc_wire_idx = len(tile_type.wire_data)
+            vcc_wire = TileWireInfo()
+            tile_type.wire_data.append(vcc_wire)
+
+            vcc_wire.name = '$VCC_WIRE'
+            vcc_wire.site = -1
+
+            tile_type_vcc_wires[tile_type_idx] = vcc_wire_idx
+
+        # Add vcc local wire in each instance to node
+        for tile_idx, tile in enumerate(self.chip_info.tiles):
+            if tile.type not in tile_types_with_vcc:
+                continue
+
+            vcc_wire_idx = tile_type_vcc_wires[tile.type]
+            assert vcc_wire_idx >= len(tile.tile_wire_to_node)
+
+            while vcc_wire_idx > len(tile.tile_wire_to_node):
+                tile.tile_wire_to_node.append(-1)
+
+            tile.tile_wire_to_node.append(vcc_node_idx)
+
+            wire_ref = TileWireRef()
+            wire_ref.tile = tile_idx
+            wire_ref.index = vcc_wire_idx
+
+            vcc_node.tile_wires.append(wire_ref)
+
+        for tile_type_idx in (tile_types_with_gnd | tile_types_with_vcc):
+            gnd_wire_idx = tile_type_gnd_wires.get(tile_type_idx, None)
+            vcc_wire_idx = tile_type_vcc_wires.get(tile_type_idx, None)
+
+            self.connect_tile_type(tile_type_idx, gnd_wire_idx, vcc_wire_idx,
+                                   bel_pins_connected_to_gnd,
+                                   bel_pins_connected_to_vcc)
+
+        # FIXME: Implement node constant sources.
+        #
+        # Node constant sources are rarer, and less important, so don't
```

Comment:
I think this'll be needed for VCC on UltraScale+ as it doesn't have tieoff bels like xc7 does

Reply:
Yes and no.  Most VCC wires in US+ were deduplicated during DeviceDatabase creatation into "Wire constant sources", see https://github.com/SymbiFlow/fpga-interchange-schema/blob/bce42208b80c239d749881603b4cdf41944c70ad/interchange/DeviceResources.capnp#L130 and https://github.com/Xilinx/RapidWright/blob/95953c12b9027cae48f6c83dc36e8e34eb33a76d/src/com/xilinx/rapidwright/interchange/ConstantDefinitions.java#L355-L380

There are handful of "strange" wires that fall into the exceptional bucket.  The VCC nodes you are thinking of are handled in the tile type wire sources!

Reply:
Device database documentation: https://github.com/SymbiFlow/fpga-interchange-schema/blob/bce42208b80c239d749881603b4cdf41944c70ad/interchange/DeviceResources.capnp#L356-L360